PC/_wmimodule.cpp

```diff
@@ -0,0 +1,302 @@
+// Licensed to PSF under a contributor agreement
+//
+
+// Version history
+//  2022-08: Initial contribution (Steve Dower)
+
+#define _WIN32_DCOM
+#include <Windows.h>
+#include <comdef.h>
+#include <Wbemidl.h>
+#include <propvarutil.h>
+
+#include <Python.h>
+#include "clinic/_wmimodule.cpp.h"
+
+
+/*[clinic input]
+module _wmi
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=7ca95dad1453d10d]*/
+
+
+
+struct _query_data {
+    LPCWSTR query;
+    HANDLE writePipe;
+    HANDLE readPipe;
+};
+
+
+static DWORD WINAPI
+_query_thread(LPVOID param)
+{
+    IWbemLocator *locator = NULL;
+    IWbemServices *services = NULL;
+    IEnumWbemClassObject* enumerator = NULL;
+    BSTR bstrQuery = NULL;
+    struct _query_data *data = (struct _query_data*)param;
+
+    HRESULT hr = CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
+    if (FAILED(hr)) {
+        CloseHandle(data->writePipe);
+        return (DWORD)hr;
+    }
+
+    hr = CoInitializeSecurity(
+        NULL, -1, NULL, NULL,
+        RPC_C_AUTHN_LEVEL_DEFAULT,
+        RPC_C_IMP_LEVEL_IMPERSONATE,
+        NULL, EOAC_NONE, NULL
+    );
+    if (SUCCEEDED(hr)) {
+        hr = CoCreateInstance(
+            CLSID_WbemLocator, 0, CLSCTX_INPROC_SERVER,
+            IID_IWbemLocator, (LPVOID *)&locator
+        );
+    }
+    if (SUCCEEDED(hr)) {
+        hr = locator->ConnectServer(
+            bstr_t(L"ROOT\\CIMV2"),
+            NULL, NULL, 0, NULL, 0, 0, &services
+        );
+    }
+    if (SUCCEEDED(hr)) {
+        hr = CoSetProxyBlanket(
+            services, RPC_C_AUTHN_WINNT, RPC_C_AUTHZ_NONE, NULL,
+            RPC_C_AUTHN_LEVEL_CALL, RPC_C_IMP_LEVEL_IMPERSONATE,
+            NULL, EOAC_NONE
+        );
+    }
+    if (SUCCEEDED(hr)) {
+        bstrQuery = SysAllocString(data->query);
+        if (!bstrQuery) {
+            hr = HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_MEMORY);
+        }
+    }
+    if (SUCCEEDED(hr)) {
+        hr = services->ExecQuery(
+            bstr_t("WQL"), 
+            bstrQuery,
+            WBEM_FLAG_FORWARD_ONLY | WBEM_FLAG_RETURN_IMMEDIATELY, 
+            NULL,
+            &enumerator
+        );
+    }
+
+    // Okay, after all that, at this stage we should have an enumerator
+    // to the query results and can start writing them to the pipe!
+    IWbemClassObject *value = NULL;
+    int startOfEnum = TRUE;
+    int endOfEnum = FALSE;
+    while (SUCCEEDED(hr) && !endOfEnum) {
+        ULONG got = 0;
+        DWORD written;
+        hr = enumerator->Next(WBEM_INFINITE, 1, &value, &got);
+        if (hr == WBEM_S_FALSE) {
+            // Could be at the end, but still got a result this time
+            endOfEnum = TRUE;
+            hr = 0;
+            break;
+        }
+        if (FAILED(hr) || got != 1 || !value) {
+            continue;
+        }
+        if (!startOfEnum && !WriteFile(data->writePipe, (LPVOID)L"\0", 2, &written, NULL)) {
+            hr = HRESULT_FROM_WIN32(GetLastError());
```

Comment:
```suggestion
            hr = HRESULT_FROM_WIN32(GetLastError());
            break;
```
If writing to the pipe fails, the enumeration has to stop.